File: server/app/schemas/artifact.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, ConfigDict


ArtifactStatus = Literal["good", "need_check", "warning", "damaged", "maintenance", "archived"]


class ArtifactBase(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=2000)
    location: str = Field(default="", max_length=200)
    status: ArtifactStatus = "good"
    inspection_interval_days: int = Field(default=0, ge=0)


class ArtifactCreate(ArtifactBase):
    # Optional schedule to be created with the artifact
    scheduled_date: datetime | None = None
    scheduled_time: str | None = Field(default="09:00", pattern=r"^\d{2}:\d{2}$")


class ArtifactUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    location: str | None = Field(default=None, max_length=200)
    status: ArtifactStatus | None = None
    inspection_interval_days: int | None = Field(default=None, ge=0)


class ArtifactRead(ArtifactBase):
    model_config = ConfigDict(from_attributes=True)

    id: str
    has_image: bool
    reference_image_path: str | None = None
    created_at: datetime
    updated_at: datetime

File: client/artifact_app/lib/models/artifact.dart
import 'artifact_status.dart';

class Artifact {
  final String id; // VARCHAR(6) from PostgreSQL
  final String name;
  final String? description;
  final String? location;
  final ArtifactStatus status;
  final int inspectionIntervalDays; // Thêm trường chu kỳ (ngày)
  final String? baselineImageId;
  final bool hasImage;
  final String? referenceImagePath;
  final DateTime createdAt;
  final DateTime updatedAt;

  Artifact({
    required this.id,
    required this.name,
    this.description,
    this.location,
    required this.status,
    required this.inspectionIntervalDays,
    this.baselineImageId,
    required this.hasImage,
    this.referenceImagePath,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Artifact.fromJson(Map<String, dynamic> json) {
    return Artifact(
      id: json['id']?.toString() ?? '',
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      location: json['location'] as String?,
      status: ArtifactStatus.fromWire(json['status'] as String?),
      inspectionIntervalDays: json['inspection_interval_days'] as int? ?? 0,
      baselineImageId: json['baseline_image_id']?.toString(),
      hasImage: json['has_image'] as bool? ?? false,
      referenceImagePath: json['reference_image_path'] as String?,
      createdAt: DateTime.tryParse(json['created_at']?.toString() ?? '') ?? DateTime.now(),
      updatedAt: DateTime.tryParse(json['updated_at']?.toString() ?? '') ?? DateTime.now(),
    );
  }
}
